server/src/main/java/org/elasticsearch/common/util/Maps.java

```diff
@@ -298,4 +298,14 @@ static int capacity(int expectedSize) {
         throw new IllegalStateException("Unsupported input format");
     }
 
+    /**
+     * This method creates a "deep" of the {@code source} map using on {@code copyValueFunction} each value
+     */
+    public static <K, V> Map<K, V> copy(Map<K, V> source, Function<V, V> copyValueFunction) {
+        var copy = Maps.<K, V>newHashMapWithExpectedSize(source.size());
+        for (var entry : source.entrySet()) {
+            copy.put(entry.getKey(), copyValueFunction.apply(entry.getValue()));
+        }
+        return copy;
```

Comment:
I wonder if it is worth implementing this change on master and then sync and fix this branch with it?